File: packages/guards/src/pii-au.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

type PiiAuEntity = 'tfn' | 'medicare-au' | 'abn' | 'passport-au' | 'phone-au';

interface PiiAuOptions { entities: PiiAuEntity[]; action: 'block' | 'warn' | 'mask'; }
interface PiiAuMatch { type: PiiAuEntity; value: string; start: number; end: number; }

const PATTERNS: Record<PiiAuEntity, RegExp> = {
  tfn: /\b\d{3}\s?\d{3}\s?\d{3}\b/g,
  'medicare-au': /\b\d{4}\s?\d{5}\s?\d{1}\b/g,
  abn: /\b\d{2}\s?\d{3}\s?\d{3}\s?\d{3}\b/g,
  'passport-au': /\b[A-Z]{1,2}\d{7}\b/g,
  'phone-au': /\b(?:\+?61|0)\s?\d\s?\d{4}\s?\d{4}\b/g,
};

const MASK_LABELS: Record<PiiAuEntity, string> = {
  tfn: '[TFN]', 'medicare-au': '[Medicare]', abn: '[ABN]',
  'passport-au': '[Passport]', 'phone-au': '[Phone]',
};

function detect(text: string, entities: PiiAuEntity[]): PiiAuMatch[] {
  const matches: PiiAuMatch[] = [];
  for (const entity of entities) {
    const re = new RegExp(PATTERNS[entity].source, 'g');
    let m: RegExpExecArray | null;
    while ((m = re.exec(text)) !== null) {
      matches.push({ type: entity, value: m[0], start: m.index, end: m.index + m[0].length });
    }
  }
  return matches.sort((a, b) => b.start - a.start);
}

function mask(text: string, matches: PiiAuMatch[]): string {
  let r = text;
  for (const m of matches) { r = r.slice(0, m.start) + MASK_LABELS[m.type] + r.slice(m.end); }
  return r;
}

export function piiAu(options: PiiAuOptions): Guard {
  return {
    name: 'pii-au', version: '0.1.0',
    description: 'Australian PII detection (TFN, Medicare, ABN, passport, phone)',
    category: 'privacy', supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matches = detect(text, options.entities);
      const triggered = matches.length > 0;
      if (!triggered) return { guardName: 'pii-au', passed: true, action: 'allow', latencyMs: Math.round(performance.now() - start) };
      if (options.action === 'mask') {
        return { guardName: 'pii-au', passed: true, action: 'override', overrideText: mask(text, matches), latencyMs: Math.round(performance.now() - start), details: { detected: matches.map(({ type, value }) => ({ type, value })) } };
      }
      return { guardName: 'pii-au', passed: false, action: options.action, message: `Australian PII: ${[...new Set(matches.map((m) => m.type))].join(', ')}`, latencyMs: Math.round(performance.now() - start), details: { detected: matches.map(({ type, value }) => ({ type, value })) } };
    },
  };
}

File: python/open_guardrail/guards/pii_au.py
"""Australian PII detection and masking guard."""

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_PATTERNS: dict[str, re.Pattern[str]] = {
    "tfn": re.compile(r"\b\d{3}\s?\d{3}\s?\d{3}\b"),
    "medicare": re.compile(r"\b\d{4}\s?\d{5}\s?\d\b"),
    "passport-au": re.compile(r"\b[A-Z]{1,2}\d{7}\b"),
    "phone-au": re.compile(r"\b(?:\+61|0061|0)\s?[2-9]\d{3}[\s\-]?\d{4}\b"),
    "abn": re.compile(r"\b\d{2}\s?\d{3}\s?\d{3}\s?\d{3}\b"),
    "driver-license-au": re.compile(r"\b\d{6,10}\b"),
}

_MASK_LABELS: dict[str, str] = {
    "tfn": "[TFN]",
    "medicare": "[MEDICARE]",
    "passport-au": "[PASSPORT-AU]",
    "phone-au": "[PHONE-AU]",
    "abn": "[ABN]",
    "driver-license-au": "[DRIVER-LICENSE-AU]",
}


def _detect(text: str, entities: list[str]) -> list[dict]:
    matches: list[dict] = []
    for entity in entities:
        pat = _PATTERNS.get(entity)
        if not pat:
            continue
        for m in pat.finditer(text):
            matches.append({"type": entity, "value": m.group(), "start": m.start(), "end": m.end()})
    matches.sort(key=lambda x: x["start"], reverse=True)
    return matches


def _mask(text: str, matches: list[dict]) -> str:
    result = text
    for m in matches:
        result = result[: m["start"]] + _MASK_LABELS.get(m["type"], "[PII]") + result[m["end"] :]
    return result


class _PiiAu:
    def __init__(self, *, action: str = "block", entities: Optional[List[str]] = None) -> None:
        self.name = "pii-au"
        self.action = action
        self._entities = entities or list(_PATTERNS.keys())

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matches = _detect(text, self._entities)
        elapsed = (time.perf_counter() - start) * 1000
        if not matches:
            return GuardResult(guard_name="pii-au", passed=True, action="allow", latency_ms=round(elapsed, 2))
        detected = [{"type": m["type"], "value": m["value"]} for m in matches]
        if self.action == "mask":
            return GuardResult(guard_name="pii-au", passed=True, action="override", override_text=_mask(text, matches), latency_ms=round(elapsed, 2), details={"detected": detected})
        return GuardResult(guard_name="pii-au", passed=False, action=self.action, message="Australian PII detected", latency_ms=round(elapsed, 2), details={"detected": detected})


def pii_au(*, action: str = "block", entities: Optional[List[str]] = None) -> _PiiAu:
    return _PiiAu(action=action, entities=entities)
